components/iosnoop/src/biosnoop.rs

```diff
@@ -0,0 +1,188 @@
+        println!("here");
+        for e in t.iter() {
+            let typ = ptr::read(e.key.as_ptr() as *const IOType);
+            let stats = ptr::read(e.value.as_ptr() as *const IOStats);
+            println!("{:?} ", typ);
+            map.insert(typ, stats);
+        }
+        Some(map)
+    } else {
+        None
+    }
+}
+
+pub struct IOContext {
+    io_stats_map: Option<HashMap<IOType, IOStats>>,
+}
+
+impl IOContext {
+    pub fn new() -> Self {
+        IOContext {
+            io_stats_map: unsafe { get_io_stats() },
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn delta(self) -> HashMap<IOType, IOStats> {
+        if let Some(prev_map) = self.io_stats_map {
+            if let Some(mut now_map) = unsafe { get_io_stats() } {
+                for (typ, stats) in prev_map {
+                    now_map.entry(typ).and_modify(|e| {
+                        println!("{} {}", stats.read, stats.write);
+                        e.read -= stats.read;
+                        e.write -= stats.write;
+                    });
+                }
+                return now_map;
+            }
+        }
+        HashMap::default()
+    }
+
+    #[allow(dead_code)]
+    pub fn delta_and_refresh(&mut self) -> HashMap<IOType, IOStats> {
+        if self.io_stats_map.is_some() {
+            if let Some(map) = unsafe { get_io_stats() } {
+                for (typ, stats) in &map {
+                    self.io_stats_map
+                        .as_mut()
+                        .unwrap()
+                        .entry(*typ)
+                        .and_modify(|e| {
+                            e.read = stats.read - e.read;
+                            e.write = stats.write - e.write;
+                        })
+                        .or_insert(stats.clone());
+                }
+
+                return self.io_stats_map.replace(map).unwrap();
+            }
+        }
+        HashMap::default()
+    }
+}
+
+pub fn init_io_snooper() -> Result<(), String> {
+    let code = include_str!("biosnoop.c");
+    let code = code.replace("##TGID##", &nix::unistd::getpid().to_string());
+    // compile the above BPF code!
+    let mut bpf = BPF::new(&code).map_err(|e| e.to_string())?;
+    // attach kprobes
+    Kprobe::new()
+        .handler("trace_pid_start")
+        .function("blk_account_io_start")
+        .attach(&mut bpf)
+        .map_err(|e| e.to_string())?;
+    Kprobe::new()
+        .handler("trace_req_completion")
+        .function("blk_account_io_completion")
+        .attach(&mut bpf)
+        .map_err(|e| e.to_string())?;
+    // the "events" table is where the "open file" events get sent
+    let stats_table = bpf.table("statsbytype").map_err(|e| e.to_string())?;
+    let type_table = bpf.table("typebypid").map_err(|e| e.to_string())?;
+    unsafe {
+        BPF_TABLE = Some((bpf, stats_table, type_table));
+    }
+    // info!("init io snooper"; "pid" => nix::unistd::getpid().to_string());
+    Ok(())
+}
+
```

Comment:
I remember there's an event loop in the first edition. Could we still keep that and manage a global IOContext in this crate protected by RWLock and updated periodically? In this way metrics flusher and limiter can share access to it.

Reply:
metrics flush and limiter can get the stats from IOContext.

Reply:
we can update it in metrics flush periodically and manage a global IOContext. No need to add an event loop.

Reply:
emmmm. We can't ensure the metrics flush and limiter consume the delta within the same period. Seems sharing access is meaningless.